url147ce07ebe_JoshuaKissoon_Kademlia_tgz-pJ8-kademlia_simulations_RoutingTableStateTestingJ8/src/kademlia/operation/ContentLookupOperation.java

```diff
@@ -0,0 +1,343 @@
+    private final ContentLookupMessage lookupMessage;
+
+    private boolean isContentFound;
+    private final SortedMap<Node, Byte> nodes;
+
+    /* Tracks messages in transit and awaiting reply */
+    private final Map<Integer, Node> messagesTransiting;
+
+    /* Used to sort nodes */
+    private final Comparator comparator;
+
+    /* Statistical information */
+    private final RouteLengthChecker routeLengthChecker;
+
+    
+    {
+        messagesTransiting = new HashMap<>();
+        isContentFound = false;
+        routeLengthChecker = new RouteLengthChecker();
+    }
+
+    /**
+     * @param server
+     * @param localNode
+     * @param params    The parameters to search for the content which we need to find
+     * @param config
+     */
+    public ContentLookupOperation(KadServer server, JKademliaNode localNode, GetParameter params, KadConfiguration config)
+    {
+        /* Construct our lookup message */
+        this.lookupMessage = new ContentLookupMessage(localNode.getNode(), params);
+
+        this.server = server;
+        this.localNode = localNode;
+        this.config = config;
+
+        /**
+         * We initialize a TreeMap to store nodes.
+         * This map will be sorted by which nodes are closest to the lookupId
+         */
+        this.comparator = new KeyComparator(params.getKey());
+        this.nodes = new TreeMap(this.comparator);
+    }
+
+    /**
+     * @throws java.io.IOException
+     * @throws kademlia.exceptions.RoutingException
+     */
+    @Override
+    public synchronized void execute() throws IOException, RoutingException
+    {
+        try
+        {
+            /* Set the local node as already asked */
+            nodes.put(this.localNode.getNode(), ASKED);
+
+            /**
+             * We add all nodes here instead of the K-Closest because there may be the case that the K-Closest are offline
+             * - The operation takes care of looking at the K-Closest.
+             */
+            List<Node> allNodes = this.localNode.getRoutingTable().getAllNodes();
+            this.addNodes(allNodes);
+            
+            /* Also add the initial set of nodes to the routeLengthChecker */
+            this.routeLengthChecker.addInitialNodes(allNodes);
+
+            /**
+             * If we haven't found the requested amount of content as yet,
+             * keey trying until config.operationTimeout() time has expired
+             */
+            int totalTimeWaited = 0;
+            int timeInterval = 10;     // We re-check every n milliseconds
+            while (totalTimeWaited < this.config.operationTimeout())
+            {
+                if (!this.askNodesorFinish() && !isContentFound)
+                {
+                    wait(timeInterval);
+                    totalTimeWaited += timeInterval;
+                }
+                else
+                {
+                    break;
+                }
+            }
+        }
+        catch (InterruptedException e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)